plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSinkHandle.java

```diff
@@ -0,0 +1,149 @@
+import io.trino.spi.QueryId;
+import io.trino.spi.exchange.ExchangeSinkHandle;
+
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class FileSystemExchangeSinkHandle
+        implements ExchangeSinkHandle
+{
+    private final QueryId queryId;
+    private final int stageId;
+    private final int taskPartitionId;
+    private final Optional<SecretKey> secretKey;
+
+    @JsonCreator
+    public FileSystemExchangeSinkHandle(
+            @JsonProperty("queryId") QueryId queryId,
+            @JsonProperty("stageId") int stageId,
+            @JsonProperty("taskPartitionId") int taskPartitionId,
+            @JsonProperty("secretKey") Optional<SecretKey> secretKey)
+    {
+        this.queryId = requireNonNull(queryId, "queryId is null");
+        this.stageId = stageId;
+        this.taskPartitionId = taskPartitionId;
+        this.secretKey = requireNonNull(secretKey, "secretKey is null");
+    }
+
+    @JsonProperty
+    public QueryId getQueryId()
+    {
+        return queryId;
+    }
+
+    @JsonProperty
+    public int getStageId()
+    {
+        return stageId;
+    }
+
+    @JsonProperty
+    public int getTaskPartitionId()
+    {
+        return taskPartitionId;
+    }
+
+    @JsonProperty
+    @JsonSerialize(using = SecretKeySerializer.class)
+    @JsonDeserialize(using = SecretKeyDeserializer.class)
+    public Optional<SecretKey> getSecretKey()
+    {
+        return secretKey;
+    }
+
+    public static class SecretKeySerializer
+            extends JsonSerializer<Optional<SecretKey>>
+    {
+        @Override
+        public void serialize(Optional<SecretKey> secretKey, JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
+                throws IOException
+        {
+            if (secretKey.isPresent()) {
+                jsonGenerator.writeString(Base64.getEncoder().encodeToString(secretKey.get().getEncoded()));
+            }
+            else {
+                jsonGenerator.writeString("");
+            }
+        }
+    }
+
+    public static class SecretKeyDeserializer
+            extends JsonDeserializer<Optional<SecretKey>>
+    {
+        @Override
+        public Optional<SecretKey> deserialize(JsonParser jsonParser, DeserializationContext deserializationContext)
+                throws IOException
+        {
+            String encodedKey = jsonParser.getValueAsString();
+            if (encodedKey.isEmpty()) {
+                return Optional.empty();
+            }
+            else {
+                byte[] decodedKey = Base64.getDecoder().decode(encodedKey);
+                return Optional.of(new SecretKeySpec(decodedKey, 0, decodedKey.length, "AES"));
+            }
+        }
+    }
+
+    @Override
+    public boolean equals(Object o)
```

Comment:
skipping key in equals/hashcode on purpose? It should not matter I guess, but curious still.

Reply:
I don't know the best practice here. But added anyways.